velox/expression/PeeledEncoding.cpp

```diff
@@ -109,6 +111,61 @@ void PeeledEncoding::setDictionaryWrapping(
   wrapNulls_ = std::move(wrapping.nulls);
 }
 
+void PeeledEncoding::flattenPeeledVectors(
+    const SelectivityVector& rows,
+    const std::vector<bool>& constantFields,
+    std::vector<VectorPtr>& peeledVectors) {
+  VELOX_CHECK(flattenPeeled_.empty());
+  for (int i = 0; i < peeledVectors.size(); ++i) {
+    if (!constantFields.empty() && constantFields[i]) {
+      continue;
+    }
+    if (peeledVectors[i]->isConstantEncoding() && !wrapNulls_) {
+      continue;
+    }
+    switch (peeledVectors[i]->typeKind()) {
+      case TypeKind::ARRAY:
+      case TypeKind::MAP:
+      case TypeKind::ROW:
+        return;
+      default:
+        if (isLazyNotLoaded(*peeledVectors[i])) {
+          return;
+        }
+    }
+  }
+  auto nonNullRows = rows;
+  if (wrapNulls_) {
+    nonNullRows.deselectNulls(
+        wrapNulls_->as<uint64_t>(), rows.begin(), rows.end());
+  }
+  for (int i = 0; i < peeledVectors.size(); ++i) {
+    if (!constantFields.empty() && constantFields[i]) {
+      continue;
+    }
+    if (peeledVectors[i]->isConstantEncoding() && !wrapNulls_) {
+      if (peeledVectors[i]->size() < rows.end()) {
+        peeledVectors[i] =
+            BaseVector::wrapInConstant(rows.end(), 0, peeledVectors[i]);
+      }
+    } else {
+      auto flat = BaseVector::create(
```

Comment:
@Yuhta NOTE: This would create a new vector every time we peel, which means if these are part of a shared subexpression then those results will never be re-used since they expect the exact same input vector to passed every time. Its possible we never see this after effect in action because the small size of the output only adds a marginal cost to expression eval. But something to keep in mind.

Reply:
Do you have some idea where we can keep these vectors so they can be reused?

I actually have another idea to avoid flattening (i.e. bail out peeling) when possible, since flattening is only needed when the underlying function/cast cannot handle dictionary input.

Reply:
Skipping peeling in that case would be a good idea, during the common expr eval code path, peeling can happen in 2 places, once at the start of eval (before CSE step) and once right before passing the input to the vectorFunction. For the first instance, we can choose to skip peeling in this case instead of flattening. what do you think?

Reply:
Yes that's what I am thinking.  Also we can add some metadata to function to indicating whether dictionary can be handled or not, then in case 2 we can also skip peeling for certain functions.